build.gradle

```diff
@@ -160,50 +211,46 @@ task zip(type: Zip) {
     }
 }
 
-task installPlugin() {
-    if (!new File("${rootDir}/dist").exists()) {
-        return
+tasks.register('installPlugin') {
+    var pluginProjects = subprojects.findAll {
+        it.file('gradle.properties').exists()
+                && it.properties.containsKey('pluginType')
+                && it.properties.containsKey('pluginName')
+    }
+    pluginProjects.
+            forEach(subProject -> {
+                var pluginType = subProject.properties.get('pluginType')
+                var pluginName = subProject.properties.get('pluginName')
+                dependsOn("${subProject.path}:jar")
```

Comment:
subProjects have already depended on `jar` in `dist` task. Is it necessary to declare it again in task `installPlugin`?

Reply:
Given [your previous comment](https://github.com/apache/eventmesh/pull/4719#discussion_r1557893664) regarding the dependance of `installPlugin` on `dist`, I am a little bit confused about the relationship between these tasks.

Is `installPlugin` always called after `dist`? If so, it should probably depend on it. On the other hand `dist` does not need to depend on `:eventmesh-connectors:eventmesh-connector-rocketmq:jar` for example, because it does not use that JAR.

What happens in the `master` branch is:

- there is a `dist` task for **each** project, even "plugin" projects,
- there is a **single** `:installPlugin` task that collects all the plugins into the `dist` folder.

Maybe it should be reversed:

- there should be a **single** `:dist` project that builds up the basic binary distribution (this is implemented in this PR),
- there should be a `installPlugin` task for **each** "plugin" project (currently in this PR, there is a single one).

This way running:
```sh
./gradlew clean :eventmesh-connectors:eventmesh-connector-rocketmq:installPlugin
```
would create a base distribution with just a single plugin. What do you think?

Reply:
> > If so, it should probably depend on it.
> 
> IMO, since `jar` tasks have already been executed in `dist` task, so they shouldn't be executed again in task `installPlugin`, just as the original `installPlugin` task didn't do that either.

You can not be sure that `dist` has already been executed, since `installPlugin` does not have a dependency on `dist`. These dependencies ensure that those JARs are ready.

If `dist` **has** been already executed. The JARs will not be built again. If you run
```
./gradlew --console plain --no-daemon clean dist installPlugin
```
nothing will be executed between the `:dist` task and `:installPlugin`. If multiple tasks in the build depend on task `foo`, the task will be executed only once (or most probably its result will be retrieved from the cache).

> > would create a base distribution with just a single plugin
> 
> That's a good idea, but I'm not sure if anyone does this though, because trying to remember the exact name of a plugin is difficult, there are a lot of plugins, and the vast majority of api's require a default plugin.
> 
> After you reminded me of this, I realized that the `installPlugin` task is actually doing the work of the `installPlugins` task. However, considering that we don't need an new `installPlugin` task, changes to the task would involve multiple changes to documentation, CI, etc., so I don't think it's really necessary.

I could do this in another PR.

Reply:
>You can not be sure that `dist` has already been executed, since `installPlugin` does not have a dependency on `dist`. These dependencies ensure that those JARs are ready.

Then `installPlugin` should also depend on the `dist` task, not the `jar` task. The `installPlugin` task won't get the full result if the `dist` task hasn't been executed.

However, if the `installPlugin` depends on the `dist` task, the user will not be able to load their own jar file, losing the flexibility of the `installPlugin` task.

Considering that both the `dist` task and the `installPlugin` task will be executed (and should be) anyway, I don't think the `installPlugin` task should depend on the `dist` task.

>If multiple tasks in the build depend on task `foo`, the task will be executed only once (or most probably its result will be retrieved from the cache).

When testing your branch on my own PC, sometimes the `installPlugin` task executes a long `compileJava` task and outputs some deprecated warnings, even though the `dist` task has already been executed. This problem may be worse on less powerful machines. So I feel that the dependency of the `installPlugin` task on the `dist` task introduces some unnecessary performance overhead.

>I could do this in another PR.

I'm don't know how I'm going to introduce this new feature in documentation ...... If I need to load a custom closed-source plugin, maybe I would manually copy the jar package to the specified location in the file explorer after having run the `dist` task, and then run the `installPlugin` task. 

A new task brings some flexibility, however another concern I have is that renaming `installPlugin` to
`installPlugins` and then adding a new `installPlugin` task that is not the same will confuse other developers in the community.

@xwm1992 What do you think?

Reply:
PMCs may be busy these days. How about we just keep `installPlugin`'s original task dependency logic for now, i.e., no dependencies on `dist` or `jar` tasks? Changes to the Gradle task logic can be addressed in another PR.

Reply:
I removed the dependencies.